File: hejoo345/week_15/PGS_전력망을 둘로 나누기.java
import java.util.*;

class Solution {
    static int[][] arr;

    public int solution(int n, int[][] wires) {
        int answer = n;
        arr = new int[n + 1][n + 1];

        for (int i = 0; i < wires.length; i++) {
            arr[wires[i][0]][wires[i][1]] = 1;
            arr[wires[i][1]][wires[i][0]] = 1;
        }

        // 전선을 하나씩 끊어봄
        for (int i = 0; i < wires.length; i++) {
            int a = wires[i][0];
            int b = wires[i][1];

            // 전선 끊기
            arr[a][b] = 0;
            arr[b][a] = 0;

            answer = Math.min(answer, bfs(n, a));

            // 전선 다시 복구
            arr[a][b] = 1;
            arr[b][a] = 1;
        }
        return answer;
    }

    public int bfs(int n, int node) {
        boolean[] visited = new boolean[n + 1];
        int cnt = 1;

        Queue<Integer> queue = new LinkedList<>();
        queue.add(node);

        while (!queue.isEmpty()) {
            int now = queue.poll();
            visited[now] = true;
            for (int i = 1; i <= n; i++) {
                if (!visited[i] && arr[now][i] == 1) {
                    queue.add(i);
                    cnt++;
                }
            }
        }
        return Math.abs(n - cnt * 2);
    }
}

File: raeunlee/week_15/PGS_전력망을둘로나누기.py
from collections import deque

def bfs(start,visited,graph):
    queue = deque([start]) #시작점
    result = 1
    visited[start] = True #방문처리
    while queue:
        now = queue.popleft() #현재지점
        
        for i in graph[now]:
            if visited[i] == False: 
                result += 1
                queue.append(i)
                visited[i] = True
                
    return result
        

def solution(n, wires):
    answer = n #초기값
    graph = [[] for _ in range(n+1)]
    
    for v1,v2 in wires:
        graph[v1].append(v2)
        graph[v2].append(v1)
            
    for start, not_visit in wires: #연결 시작, 끝 점
        visited = [False]*(n+1)
        visited[not_visit] = True 
        result = bfs(start,visited,graph) #해당 선을 끊어서 그 결과를 result에 저장
        if abs(result - (n-result)) < answer: #result와 전체송전탑-result의 차이계산
            answer = abs(result - (n-result)) 
        
    return answer